File: frontend/src/lib/resume.ts
import { apiRequest } from "./api";

export type ContactInfo = {
  name: string | null;
  email: string | null;
  phone: string | null;
  linkedin: string | null;
  github: string | null;
};

export type ParsedResume = {
  contact: ContactInfo;
  skills: Record<string, string[]>;
  education: string[];
  experience: string[];
  internships: string[];
  projects: string[];
  certifications: string[];
  achievements: string[];
  summary: string | null;
};

export type ResumeUploadResult = {
  filename: string;
  file_type: "pdf" | "docx";
  character_count: number;
  word_count: number;
  extracted_text: string;
  parsed: ParsedResume;
};

export const SKILL_CATEGORY_LABELS: Record<string, string> = {
  programming: "Programming Languages",
  frontend: "Frontend",
  backend: "Backend",
  database: "Database",
  tools: "Tools",
};

export const ACCEPTED_RESUME_TYPES = [
  "application/pdf",
  "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
];
export const ACCEPTED_RESUME_EXTENSIONS = [".pdf", ".docx"];
export const MAX_RESUME_SIZE_MB = 5;

export function isAcceptedResumeFile(file: File): boolean {
  if (ACCEPTED_RESUME_TYPES.includes(file.type)) return true;
  const lowerName = file.name.toLowerCase();
  return ACCEPTED_RESUME_EXTENSIONS.some((ext) => lowerName.endsWith(ext));
}

export async function uploadResume(file: File): Promise<ResumeUploadResult> {
  const formData = new FormData();
  formData.append("file", file);
  return apiRequest<ResumeUploadResult>("/resume/upload", {
    method: "POST",
    body: formData,
  });
}

File: backend/app/schemas/resume.py
"""Pydantic request/response schemas for resume upload and parsing."""
from typing import Optional

from pydantic import BaseModel, Field

# Generous upper bound for pasted/extracted resume text — far beyond any real
# resume, but bounds the payload size reaching regex/AI processing downstream.
MAX_RESUME_TEXT_LENGTH = 50_000


class ContactInfo(BaseModel):
    name: Optional[str] = None
    email: Optional[str] = None
    phone: Optional[str] = None
    linkedin: Optional[str] = None
    github: Optional[str] = None


class ParsedResume(BaseModel):
    contact: ContactInfo
    skills: dict[str, list[str]]
    education: list[str]
    experience: list[str]
    internships: list[str]
    projects: list[str]
    certifications: list[str]
    achievements: list[str]
    summary: Optional[str] = None


class ResumeUploadResponse(BaseModel):
    filename: str
    file_type: str  # "pdf" | "docx"
    character_count: int
    word_count: int
    extracted_text: str
    parsed: ParsedResume


class ATSScoreRequest(BaseModel):
    """Re-analyze already-extracted text (from a prior /upload call) — avoids
    re-uploading the file just to compute a score."""
    extracted_text: str = Field(min_length=1, max_length=MAX_RESUME_TEXT_LENGTH)


class ATSScoreResponse(BaseModel):
    overall_score: int
    keyword_score: int
    skills_score: int
    structure_score: int
    experience_score: int
    project_score: int
    formatting_score: int
    parsed: ParsedResume
